ops/main.py

```diff
@@ -42,6 +43,134 @@
+
+    # The action's uuid, for example '1' (from JUJU_ACTION_UUID).
+    action_uuid: Optional[str] = None
+
+    # The root directory of the charm where it is running,
+    # for example '/var/lib/juju/agents/unit-bare-0/charm' (from JUJU_CHARM_DIR).
+    # If JUJU_CHARM_DIR is None or set to an empty string,
+    # use Path(f'{__file__}/../../..') as default
+    # (assuming the '$JUJU_CHARM_DIR/lib/op/main.py' structure).
+    charm_dir: Path = Path()
+
+    # If true, write logs to stderr as well as to juju-log (from JUJU_DEBUG).
+    debug: bool = False
+
+    # In the format of 'actions/do-something',
+    # for example 'hooks/workload-pebble-ready' (from JUJU_DISPATCH_PATH).
+    dispatch_path: str = ''
+
+    # The name of the model, for example 'foo' (from JUJU_MODEL_NAME).
+    model_name: str = ''
+
+    # The uuid of the model,
+    # for example 'cdac5656-2423-4388-8f30-41854b4cca7d' (from JUJU_MODEL_UUID).
+    model_uuid: str = ''
+
+    # The ID of the notice, for example '1', (from JUJU_NOTICE_ID).
+    notice_id: Optional[str] = None
+
+    # The key of the notice, for example 'example.com/a', (from JUJU_NOTICE_KEY).
+    notice_key: Optional[str] = None
+
+    # The type of the notice, for example 'custom' (from JUJU_NOTICE_TYPE).
+    notice_type: Optional[str] = None
+
+    # The name of the pebble check, for example 'http-check' (from JUJU_PEBBLE_CHECK_NAME).
+    pebble_check_name: Optional[str] = None
+
+    # The unit that is departing a relation, for example 'remote/42' (from JUJU_DEPARTING_UNIT).
+    relation_departing_unit_name: Optional[str] = None
+
+    # The name of the relation, for example 'database' (from JUJU_RELATION).
+    relation_name: Optional[str] = None
+
+    # The id of the relation,
+    # for example 1 (integer) if the original environment variable's value is 'database:1'
+    # (from JUJU_RELATION_ID).
+    relation_id: Optional[int] = None
+
+    # The name of the remote app, for example 'remoteapp1' (from JUJU_REMOTE_APP).
+    remote_app: Optional[str] = None
+
+    # The name of the remote unit, for example 'remoteapp1/0' (from JUJU_REMOTE_UNIT).
+    remote_unit_name: Optional[str] = None
+
+    # The ID of the secret,
+    # for example 'secret:dcc7aa9c-7202-4da6-8d5f-0fbbaa4e1a41' (from JUJU_SECRET_ID).
+    secret_id: Optional[str] = None
+
+    # The label of the secret, for example 'db-password' (from JUJU_SECRET_LABEL).
+    secret_label: Optional[str] = None
+
+    # The revision of the secret, for example 42 (integer) (from JUJU_SECRET_REVISION).
+    secret_revision: Optional[int] = None
+
+    # The storage name, for example 'my-storage'
+    # if the original environment variable's value is 'my-storage/1' (from JUJU_STORAGE_ID).
+    storage_name: Optional[str] = None
+
+    # The name of the unit, for example 'myapp/0' (from JUJU_UNIT_NAME)
+    unit_name: str = ''
+
+    # The version of Juju, for example '3.4.0' (from JUJU_VERSION)
+    version: Optional[JujuVersion] = None
```

Comment:
No, it can't be None; I only make it optional with a default None value so that it can be put in the right place alphabetically. And if I give it a default value like `JujuVersion('0.0.0')`, ruff would complain about it. Shall I make it non-optional without default value and put it as the first attribute of the class?

Reply:
You should definitely be able to do:

```suggestion
    version: JujuVersion = dataclasses.field(default_factory=lambda: JujuVersion('0.0.0'))
```

I still don't love the "0.0.0" version, but I suppose it's ok as a "doesn't have any of the new features" default, and probably better than forcing people to provide a version every time, when most of the time they probably don't care.